backend/src/apiserver/server/pipeline_server.go

```diff
@@ -98,427 +108,1027 @@ type PipelineServer struct {
+	// Fail if pipeline spec is missing
+	if (pv.PipelineSpec == "") && (pv.CodeSourceUrl == "") && (pv.PipelineSpecURI == "") {
+		return nil, util.NewInvalidInputError("[PipelineServer %s]: Failed to create a pipeline version due to missing pipeline spec.", s.options.ApiVersion)
 	}
+
+	// Check authorization
 	resourceAttributes := &authorizationv1.ResourceAttributes{
-		Verb: common.RbacResourceVerbUpdate,
+		Verb: common.RbacResourceVerbList,
 	}
-	err := s.CanAccessPipeline(ctx, request.PipelineId, resourceAttributes)
-	if err != nil {
-		return nil, util.Wrap(err, "Failed to authorize the requests.")
+	if err := s.canAccessPipeline(ctx, pv.PipelineId, resourceAttributes); err != nil {
+		return nil, util.Wrap(err, fmt.Sprintf("[PipelineServer %s]: Failed to create a pipeline version due authorization error for pipeline id %v.", s.options.ApiVersion, pv.PipelineId))
 	}
-	err = s.resourceManager.UpdatePipelineDefaultVersion(request.PipelineId, request.VersionId)
-	if err != nil {
-		return nil, util.Wrap(err, "Update Pipeline Default Version failed.")
+
+	// Read pipeline file
+	if pv.PipelineSpec == "" {
+		pipelineUrl, err := url.ParseRequestURI(pv.CodeSourceUrl)
+		if err != nil {
+			pipelineUrl, err = url.ParseRequestURI(pv.PipelineSpecURI)
+			if err != nil {
+				return nil, util.NewInvalidInputError("[PipelineServer %s]: Failed to create a pipeline version due to invalid pipeline spec URI. PipelineSpecURI: %v. CodeSourceUrl: %v. Please specify a valid URL", s.options.ApiVersion, pv.PipelineSpecURI, pv.CodeSourceUrl)
+			}
+		}
+		resp, err := s.httpClient.Get(pipelineUrl.String())
+		if err != nil || resp.StatusCode != http.StatusOK {
+			return nil, util.NewInternalServerError(err, "[PipelineServer %s]: Failed to create a pipeline version due error downloading the pipeline spec from %v.", s.options.ApiVersion, pipelineUrl.String())
+		}
+		pipelineFileName := path.Base(pipelineUrl.String())
+		pipelineFile, err := ReadPipelineFile(pipelineFileName, resp.Body, common.MaxFileLength)
+		if err != nil {
+			return nil, util.Wrap(err, fmt.Sprintf("[PipelineServer %s]: Failed to create a pipeline version due error reading the pipeline spec.", s.options.ApiVersion))
+		}
+		pv.PipelineSpec = string(pipelineFile)
+		if pv.Name == "" {
+			pv.Name = pipelineFileName
+		}
 	}
-	return &empty.Empty{}, nil
+
+	// Validate the pipeline version
+	if err := s.validatePipelineVersionBeforeCreating(&pv); err != nil {
+		return nil, util.Wrap(err, fmt.Sprintf("[PipelineServer %s]: Failed to create a pipeline version due to validation error.", s.options.ApiVersion))
+	}
+
+	return s.resourceManager.CreatePipelineVersion(pv)
 }
 
-func (s *PipelineServer) GetPipelineV1(ctx context.Context, request *api.GetPipelineRequest) (*api.Pipeline, error) {
-	if s.options.CollectMetrics {
-		getPipelineRequests.Inc()
+// Fetches a model.Pipeline.
+func (s *PipelineServer) getPipeline(ctx context.Context, pipelineId string) (*model.Pipeline, error) {
```

Comment:
Since this method should be accessible from other packages, it should start with an upper case.

Reply:
This function is not called outside of `PipelineServer`